### PR TITLE
Improve annotation of `fromstring`

### LINKED ISSUE #63: return type for `lxml.etree.fromstring` should be Optional (?)
At least with lxml 4.8.0, I see the following:

```
$ python
Python 3.10.4 (main, Mar 25 2022, 00:00:00) [GCC 11.2.1 20220127 (Red Hat 11.2.1-9)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import lxml.etree as etree
>>> parser = etree.HTMLParser(recover=True, encoding="utf-8")
>>> tree = etree.fromstring(b"", parser)
>>> tree
>>> tree is None
True
```

But the type annotation says we'll always return an `_Element`:

https://github.com/lxml/lxml-stubs/blob/5eaa0f8723da181192ca3e20b478ebb89588228b/lxml-stubs/etree.pyi#L498-L500

Aside: I'm passing an HTMLParser here instead of an XMLParser. The type stub for `fromstring` doesn't allow that either, but again I'm not sure if that's a mistake in the annotations.

For context, I came across this whilst trying to use the stubs to typecheck [this snippet of Synapse](https://github.com/matrix-org/synapse/blob/13e4386710243860ccf1b6b8127c7a3b89df2e2c/synapse/rest/media/v1/oembed.py#L236-L241).

### PR DESCRIPTION
**Pull request by DMRobertson**
Resolves #63. See discussion there for more context.

1. `parser` seems like it can be an `HTMLParser`. There's code in the wild
that does this without obviously being wrong. As far as I can tell from
the lxml Cython source, `parser` is a `_BaseParser`, which encompasses
`XMLParser`, `HTMLParser` and the `iterparse` class. I wasn't sure if the 
latter  ought to be allowed here too. I erred on the side of caution.

2. It's possible for this function to return None, if the parser has
`recover=True`. I couldn't find a good way to express this without
affecting every other call to `fromstring`. (Perhaps one could make the
Parser generic over some kind of `Recoverable` indicator type... but
that seemed like overkill).

### CODE REVIEW

File: lxml-stubs/etree.pyi [hunk 1]
```diff
@@ -496,8 +496,11 @@ def parse(
     source: _FileSource, parser: XMLParser = ..., base_url: _AnyStr = ...
```
Comment (scoder) [inline lxml-stubs/etree.pyi:496]: The `HTMLParser` is also allowed here.

File: test-data/test-etree.yml [hunk 2]
```diff
@@ -3,7 +3,15 @@
     main: |
         from lxml import etree
         document = etree.fromstring("<doc></doc>")
-        reveal_type(document)  # N: Revealed type is "lxml.etree._Element"
+        reveal_type(document)  # N: Revealed type is "Union[lxml.etree._Element, None]"
```
Comment (scoder) [inline test-data/test-etree.yml:6]: Since `None` is really a thing that you'd only get if you pass a suitable parser (and that gets in the way otherwise), maybe we can split up the declarations of `fromstring()` and `parse()` to only allow `None` return values if the parser argument is provided?

Admittedly, you can also change the default parser globally, which then allows a `None` return value here. And, in fact, it's not even guaranteed that the return value is an `Element` otherwise, since parsers can really return whatever they choose. However, that's such an extremely rare use case that I'd lean towards not expressing it in the type system. (OTOH, the whole point of this PR is to cater for an extremely rare use case, so …)

Comment (JelleZijlstra) [reply test-data/test-etree.yml:6]: In typeshed we generally avoid Unions in return types, for reasons discussed in https://github.com/python/mypy/issues/1693. In this case, it sounds like returning None is a very rare edge case, but if we return `Element | None`, we force every caller to check for None after calling this function. As you say, one improvement could be to use overloads for cases where we know the return value is definitely not None. Another approach we've used in typeshed is to return something like `Element | Any`. This gives you a lot of type safety but also allows callers to do None checks without hitting "unreachable code" warnings.

Comment (DMRobertson) [reply test-data/test-etree.yml:6]: Many thanks both!

On returning Union types: I'd read some of the discussion in https://github.com/python/typing/issues/566 which sounded promising at first, but as Jukka writes [here](https://github.com/python/typing/issues/566#issuecomment-399403480):

> For new code and new APIs the recommended way is to avoid signatures that would require the use of AnyOf anyway.

I'm about to push three commits which:
- clarify that `parse` accepts an HTML parser too;
- express that `fromstring()` without a parser returns an `_Element` (the comment about changing the default parser notwithstanding; and
- change the return type of `fromstring` to `_Element | Any`.

To be explicit: I don't have strong opinions about this PR. I like the idea of annotations that describe everything that a function can possibly do; but there's a balance to between that and pragmatism, and I'm happy to drop the `| None` change if it's too niche. It'd be nice to get the bit about HTMLParser in though, if nothing else!

File: lxml-stubs/etree.pyi [hunk 3]
```diff
@@ -493,12 +493,25 @@ def cleanup_namespaces(
     keep_ns_prefixes: Optional[Iterable[_AnyStr]] = ...,
 ) -> None: ...
 def parse(
-    source: _FileSource, parser: XMLParser = ..., base_url: _AnyStr = ...
+    source: _FileSource,
+    parser: Union[XMLParser, HTMLParser] = ...,
+    base_url: _AnyStr = ...,
 ) -> _ElementTree: ...
```
Comment (scoder) [inline lxml-stubs/etree.pyi:499]: `parse()` has the same behaviour as `fromstring()`, the only difference is the type of input. It can also return "anything".
(BTW, "Any" is really correct here. A parser `target` can build and return whatever it wants – but that requires an appropriately configured parser argument.)

```suggestion
) -> Union[_ElementTree, Any]: ...
```

**Comment by scoder** [comment]
Thanks
